packages-exp/auth-exp/src/platform_cordova/popup_redirect.test.ts

```diff
@@ -0,0 +1,131 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as chaiAsPromised from 'chai-as-promised';
+import * as sinonChai from 'sinon-chai';
+import { expect, use } from 'chai';
+import { testAuth, TestAuth } from '../../test/helpers/mock_auth';
+import { SingletonInstantiator } from '../core/util/instantiator';
+import { AuthEventType, PopupRedirectResolver } from '../model/popup_redirect';
+import { cordovaPopupRedirectResolver } from './popup_redirect';
+import { GoogleAuthProvider } from '../core/providers/google';
+import { FirebaseError } from '@firebase/util';
+
+use(chaiAsPromised);
+use(sinonChai);
+
+describe('platform_cordova/popup_redirect', () => {
+  let auth: TestAuth;
+  let resolver: PopupRedirectResolver;
+
+  beforeEach(async () => {
+    auth = await testAuth();
+    resolver = new (cordovaPopupRedirectResolver as SingletonInstantiator<PopupRedirectResolver>)();
+  });
+
+  describe('_openRedirect plugin checks', () => {
+    // TODO: Rest of the tests go here
+    it('does not reject if all plugins installed', () => {
+      setExpectedPluginsButMissing([]);
+      expect(() =>
+        resolver._openRedirect(
+          auth,
+          new GoogleAuthProvider(),
+          AuthEventType.SIGN_IN_VIA_REDIRECT
+        )
+      ).not.to.throw;
+    });
+
+    it('rejects if universal links is missing', () => {
+      setExpectedPluginsButMissing(['universalLinks.subscribe']);
+      expect(() =>
+        resolver._openRedirect(
+          auth,
+          new GoogleAuthProvider(),
+          AuthEventType.SIGN_IN_VIA_REDIRECT
+        )
+      ).to.throw(FirebaseError, 'auth/invalid-cordova-configuration');
+    });
+
+    it('rejects if build info is missing', () => {
+      setExpectedPluginsButMissing(['BuildInfo.packageName']);
+      expect(() =>
+        resolver._openRedirect(
+          auth,
+          new GoogleAuthProvider(),
+          AuthEventType.SIGN_IN_VIA_REDIRECT
+        )
+      ).to.throw(FirebaseError, 'auth/invalid-cordova-configuration');
+    });
+
+    it('rejects if browsertab openUrl', () => {
+      setExpectedPluginsButMissing(['cordova.plugins.browsertab.openUrl']);
+      expect(() =>
+        resolver._openRedirect(
+          auth,
+          new GoogleAuthProvider(),
+          AuthEventType.SIGN_IN_VIA_REDIRECT
+        )
+      ).to.throw(FirebaseError, 'auth/invalid-cordova-configuration');
+    });
+
+    it('rejects if InAppBrowser is missing', () => {
+      setExpectedPluginsButMissing(['cordova.InAppBrowser.open']);
+      expect(() =>
+        resolver._openRedirect(
+          auth,
+          new GoogleAuthProvider(),
+          AuthEventType.SIGN_IN_VIA_REDIRECT
+        )
+      ).to.throw(FirebaseError, 'auth/invalid-cordova-configuration');
+    });
+  });
+});
+
+function setExpectedPluginsButMissing(missingPlugins: string[]): void {
```

Comment:
IMO it's probably easier to just have this function mock the whole thing (with all plugins), and then write things like `delete window.universalLinks.subscribe` in individual tests. It's more straightforward and don't need the awkward object traversing below.

Reply:
Done. Still have to do a bit of a dance because the package doesn't mark those as optional. Still cleaner :+1: